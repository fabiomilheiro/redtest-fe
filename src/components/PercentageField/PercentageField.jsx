import React from "react";
import PropTypes from "prop-types";
import { Field } from "react-final-form";
import {
  FormControl,
  InputLabel,
  Input,
  FormHelperText,
  InputAdornment,
} from "@material-ui/core";
import composeValidators from "../../validators/composeValidators";
import percentage from "./../../validators/percentage";

const PercentageField = ({ name, validators, label }) => {
  const id = name;
  const helperTextId = `${id}-helper-text`;
  return (
    <Field
      name={name}
      validate={composeValidators([...validators, percentage])}
    >
      {({ input, meta }) => (
        <FormControl error={meta.error && meta.touched}>
          <InputLabel htmlFor={id}>{label}</InputLabel>
          <Input
            id={id}
            aria-describedby={label}
            endAdornment={<InputAdornment position="end">%</InputAdornment>}
            inputProps={{ "data-testid": id }}
            {...input}
          />
          <FormHelperText id={helperTextId} data-testid={helperTextId}>
            {meta.error && meta.touched
              ? meta.error
              : "A number between 0 and 100."}
          </FormHelperText>
        </FormControl>
      )}
    </Field>
  );
};

PercentageField.propTypes = {
  name: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  validators: PropTypes.arrayOf(PropTypes.func).isRequired,
};

export default PercentageField;
