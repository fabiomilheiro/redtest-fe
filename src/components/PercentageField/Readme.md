Use this component in order to place a percentage field in the form.

The percentage validator is already included, so you don't need to pass it.

```jsx
import { Form } from "react-final-form";
import required from "../../validators/required";
<Form
  onSubmit={(values) => {}}
  validate={() => {}}
  render={({ handleSubmit }) => (
    <form onSubmit={handleSubmit}>
      <PercentageField
        name="SomeName"
        label="Some name"
        validators={[required]}
      />
    </form>
  )}
/>;
```
