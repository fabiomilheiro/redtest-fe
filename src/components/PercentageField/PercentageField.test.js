import React from "react";
import { render, fireEvent } from "@testing-library/react";
import PercentageField from "./PercentageField";
import { Form } from "react-final-form";

const renderPercentageField = (input) => {
  input.label = input.label || input.name;
  input.validators = input.validators || [];

  return render(
    <Form
      onSubmit={(values) => {}}
      validate={() => {}}
      render={({ handleSubmit }) => (
        <form onSubmit={handleSubmit}>
          <PercentageField
            name={input.name}
            label={input.label}
            validators={input.validators}
          />
        </form>
      )}
    />
  );
};

const changeText = (component, value) => {
  const input = component.container.getElementsByTagName("input")[0];
  fireEvent.change(input, { target: { value: value } });
  fireEvent.blur(input);
};

const assertHelperText = (component, name, text) => {
  const helperText = component.getByTestId(`${name}-helper-text`);
  expect(helperText.textContent).toEqual(text);
};

it("renders the percentage field", () => {
  const sut = renderPercentageField({ name: "x", label: "Some name" });

  sut.getByText(/Some name/);
  expect(sut).toMatchSnapshot();
});

it("renders error for negative values", () => {
  const input = { name: "x" };
  const sut = renderPercentageField(input);
  changeText(sut, "-1");

  assertHelperText(
    sut,
    input.name,
    "The field requires a percentage value between 0 and 100."
  );
});

it("renders error for values over 100", () => {
  const input = { name: "x" };
  const sut = renderPercentageField(input);
  changeText(sut, "101");

  assertHelperText(
    sut,
    input.name,
    "The field requires a percentage value between 0 and 100."
  );
});

it("renders error for non-numeric characters", () => {
  const input = { name: "x" };
  const sut = renderPercentageField(input);
  changeText(sut, "arnold");

  assertHelperText(
    sut,
    input.name,
    "The field requires a percentage value between 0 and 100."
  );
});

it("renders custom validator error message, if not valid", () => {
  const mustBeDifferentThan11 = (value) =>
    value === "11" ? "11 not allowed." : undefined;

  const input = { name: "x", validators: [mustBeDifferentThan11] };
  const sut = renderPercentageField(input);

  changeText(sut, "11");

  assertHelperText(sut, input.name, "11 not allowed.");
});
