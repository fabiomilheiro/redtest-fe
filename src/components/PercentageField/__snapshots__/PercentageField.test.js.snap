// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the percentage field 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form>
        <div
          class="MuiFormControl-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
            data-shrink="false"
            for="x"
          >
            Some name
          </label>
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd"
          >
            <input
              aria-describedby="Some name"
              aria-invalid="false"
              class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedEnd"
              data-testid="x"
              id="x"
              name="x"
              type="text"
              value=""
            />
            <div
              class="MuiInputAdornment-root MuiInputAdornment-positionEnd"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
              >
                %
              </p>
            </div>
          </div>
          <p
            class="MuiFormHelperText-root"
            data-testid="x-helper-text"
            id="x-helper-text"
          >
            A number between 0 and 100.
          </p>
        </div>
      </form>
    </div>
  </body>,
  "container": <div>
    <form>
      <div
        class="MuiFormControl-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
          data-shrink="false"
          for="x"
        >
          Some name
        </label>
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd"
        >
          <input
            aria-describedby="Some name"
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedEnd"
            data-testid="x"
            id="x"
            name="x"
            type="text"
            value=""
          />
          <div
            class="MuiInputAdornment-root MuiInputAdornment-positionEnd"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
            >
              %
            </p>
          </div>
        </div>
        <p
          class="MuiFormHelperText-root"
          data-testid="x-helper-text"
          id="x-helper-text"
        >
          A number between 0 and 100.
        </p>
      </div>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
